### ISSUE TITLE
Security enforcement

### ISSUE THREAD

**Issue by akademic** [nixys/nxs-data-anonymizer#3]
Introduced features are great!

But database structure evolves and you can't just setup filters once and live in peace.

One day someone will add new table with secrets and theese secrets will go to staging without obfuscation.

It will be great to have option to skip all tables not added to config.
Also this option will obfuscate all undescribed columns.

This change will force to update nxs-data-anonymizer config every time we have changed database structure (added column, created table).
This will provide "secure by default" mode.

**Comment by borisershov** [comment]
You mean something like this (example):

```yaml
global:
  columns:
    api:
      value: "{{- randAlphaNum 50 | nospace | lower -}}"
filters:
  groups:
    columns:
      title:
        value: "{{- randAlphaNum 50 | nospace | lower -}}"
``` 

Globally a filters will be applied for column `api` for every table within the database. And for column `title` only for database `groups`.

?

**Comment by akademic** [comment]
Something like:

```
global:
   skip_tables_without_filters: true
   randomize_columns_without_filters: true

filters:
  groups:
    columns:
      id:
        passthrough: true
      api:
        value: "{{- randAlphaNum 50 | nospace | lower -}}"
      // other columns if exist will be randomized with respect to types: int -> random int, string ->random string
```

**Comment by akademic** [comment]
In product you referred to they have syntax:

```
filters
  only:
```
https://github.com/datanymizer/datanymizer#tables-filter

But I have not found option for default hiding undescribed in filters columns

**Comment by borisershov** [comment]
Ok, I see. Let me think about this a couple of days )

**Comment by escalate** [comment]
This feature would be very helpful.

**Comment by borisershov** [comment]
Hi everyone! I've summarized your suggestions. Please check the config options described below (draft). Does it solves the issue?

# Global settings

| Option         | Type   | Required | Default value | Description                                                      |
|---             | :---:  | :---:    | :---:         |---                                                               |
| `security`      | [Security](#security-settings) | No       | -      | Security enforcement for anonymizer |


## Security settings

| Option         | Type   | Required | Default value | Description                                                      |
|---             | :---:  | :---:    | :---:         |---                                                               |
| `policy`      | [Policy](#policy-settings) | No       | -      | Security policy for entities |
| `exceptions`      | [Exceptions](exceptions-settings) | No       | -      | Exceptions for entities |
| `defaults`      | [Defaults](defaults-settings) | No       | -      | Default filters for entities |

### Policy settings

| Option         | Type   | Required | Default value | Description                                                      |
|---             | :---:  | :---:    | :---:         |---                                                               |
| `tables`      | String | No       | -      | Default security policy for tables. If value `skip` is used all undescribed tables in config will be skipped while anonymization |
| `columns`      | String | No       | -      | Default security policy for columns. If value `randomize` is used all undescribed columns in config will be randomized (with respect to types) while anonymization |

### Exceptions settings

| Option         | Type   | Required | Default value | Description                                                      |
|---             | :---:  | :---:    | :---:         |---                                                               |
| `tables`      | Slice of strings | No       | -      | Table names without filters which are not be skipped while anonymization if option `security.policy.tables` set to `skip` |
| `columns`      | Slice of strings | No       | -      | Column names (in any table) without filters which are not be randomized while anonymization if option `security.policy.columns` set to `randomize` |

### Defaults settings

| Option         | Type   | Required | Default value | Description                                                      |
|---             | :---:  | :---:    | :---:         |---                                                               |
| `columns`      | Map of Filters | No       | -      | Default filter for columns (in any table). That filters will be applied for columns with this names without described filters |


Example of config:

```yaml
security:
  policy:
    tables: skip
    columns: randomize
  exceptions:
    tables:
    - goods
    columns:
    - id
  defaults:
    columns:
      password:
        value: "{{- randAlphaNum 50 | nospace | lower -}}"

filters:
  ...

**Comment by escalate** [comment]
Hello @borisershov ,

I don't understand that point.

"Default security policy for tables. If value skip is used all undescribed tables in config will be skipped while anonymization"

Does "skipped" mean that the table is preserved or does "skipped" mean that the table is excluded and is not part of the output?

Best regards
Felix

**Comment by borisershov** [comment]
Hi, @escalate!

> Does "skipped" mean that the table is preserved or does "skipped" mean that the table is excluded and is not part of the output?

The second, i.e. the tables will be dropped from a resulting anonymized dump.

A logic of a new `security` section is following:

- For **tables** that have any rules in a `filters` section:
  - For **columns** that have any rules in `filters` section: will be anonymized in appropriate with specified rules
  - For **columns** without any rules in `filters` section:
    - If **column** name specified in `security.exceptions.columns`: passed without any changes
    - If not:
      - If **column** name specified in `security.defaults.columns`: will be anonymized in appropriate with specified rules
      - For all other **columns**: randomize its values with respect to types
- For **tables** without any rules in `filters` section:
  - If **table** name specified in `security.exceptions.tables`: passed without any changes 
  - If not: dropped from a resulting anonymized dump (table is excluded and is not part of the output)

**Comment by borisershov** [comment]
Does it that you are expect from anonymizer security enforcement?

**Comment by tsyganov-ivan** [comment]
Hi!
We also interested in this feature. 
Latest logic looks brilliant for our case.

**Comment by borisershov** [comment]
Hi everyone! Features described above has released. At the time new abilities available only for MySQL (for PgSQL will be added in next release)

**Comment by borisershov** [comment]
Hi all! Requested abilities released in [v1.6.0](https://github.com/nixys/nxs-data-anonymizer/releases/tag/v1.6.0)
